src/algorithm/proj.rs

```diff
@@ -0,0 +1,423 @@
+//! and vice versa (inverse projection), while conversion is intended for transformations and conversions
+//! between projected coordinate systems.
+//! From the PROJ [documentation](http://proj4.org/operations/index.html):
+//!
+//! - At the most generic level, a coordinate operation is a change
+//! of coordinates, based on a one-to-one relationship, from one
+//! coordinate reference system to another.
+//! - A **transformation** is a coordinate operation in which the two
+//! coordinate reference systems are based on different datums, e.g.
+//! a change from a global reference frame to a regional frame.
+//! - A **conversion** is a coordinate operation in which both
+//! coordinate reference systems are based on the same datum,
+//! e.g. change of units of coordinates.
+//! - A **projection** is a coordinate conversion from an _ellipsoidal_
+//! coordinate system to a _plane_. Although projections are simply
+//! conversions according to the standard, they are treated as
+//! separate entities in PROJ.
+//!
+//! The `Project` and `Convert` traits are available to any `Geometry` implementing `MapCoords`.
+use std::marker::Sized;
+use proj_sys::proj_errno;
+use libc::{c_char, c_double, c_int};
+use std::ffi::CString;
+use types::{CoordinateType, Point};
+use algorithm::map_coords::MapCoordsFallible;
+use std::ffi::CStr;
+use std::str;
+use failure::Error;
+use proj_sys::{pj_strerrno, proj_context_create, proj_create, proj_create_crs_to_crs,
+               proj_destroy, proj_pj_info, proj_trans, PJconsts, PJ_AREA, PJ_COORD,
+               PJ_DIRECTION_PJ_FWD, PJ_DIRECTION_PJ_INV, PJ_LP, PJ_XY};
+
+/// Easily get a String from the external library
+fn _string(raw_ptr: *const c_char) -> String {
+    let c_str = unsafe { CStr::from_ptr(raw_ptr) };
+    str::from_utf8(c_str.to_bytes()).unwrap().to_string()
+}
+
+/// Look up an error message using the error code
+fn error_message(code: c_int) -> String {
+    let rv = unsafe { pj_strerrno(code) };
+    return _string(rv);
+}
+
+/// A `PROJ` projection or transformation instance
+pub struct Proj {
+    c_proj: *mut PJconsts,
+}
+
+impl Proj {
+    /// Try to instantiate a new `PROJ` instance
+    ///
+    /// **Note:** for projection operations, `definition` specifies
+    /// the **output** projection; input coordinates
+    /// are assumed to be geodetic in radians, unless an inverse projection is intended.
+    ///
+    /// For conversion operations, `definition` defines input, output, and
+    /// any intermediate steps that are required. See the `convert` example for more details.
+
+    // In contrast to PROJ v4.x, the type of transformation
+    // is signalled by the choice of enum used as input to the PJ_COORD union
+    // PJ_LP signals projection of geodetic coordinates, with output being PJ_XY
+    // and vice versa, or using PJ_XY for conversion operations
+    pub fn new(definition: &str) -> Option<Proj> {
```

Comment:
Having never used PROJ directly, what's an example `str` that one would pass in here? WKT of the projection?

Reply:
Nevermind--I just realized that's a dumb question. 😬